nautobot_circuit_maintenance/views.py

```diff
@@ -1,20 +1,226 @@
+
+        Args:
+            start_date (datetime.date): Date to start the search.
+
+        Returns:
+            dict: A dictionary that represents the historical matrix maintenance record data.
+
+            {
+                "7 Days": count of past 7 days of maintenance,
+                "30 Days": count of past 30 days of maintenance,
+                "365 Days": count of past 30 days of maintenances
+            }
+        """
+        # TODO: Move to a generic function set up, since this is something that should be exposed via the Capacity
+        #       Metrics plugin when enabled.
+        return_dict = {
+            "past_7_days_maintenance": self.get_maintenance_past_n_days(start_date=start_date, n_days=-7),
+            "past_30_days_maintenance": self.get_maintenance_past_n_days(start_date=start_date, n_days=-30),
+            "past_365_days_maintenance": self.get_maintenance_past_n_days(start_date=start_date, n_days=-365),
+        }
+        return return_dict
+
+    def calculate_future_maintenances(self, start_date: datetime.date):
+        """Method to calculate future maintenances.
+
+        Args:
+            start_date (datetime.date): Date to start the search.
+
+        Returns:
+            int: Count of future maintenances
+        """
+        count = 0
+        for ckt_maint in self.queryset.all():
+            if ckt_maint.start_time.date() > start_date:
+                count += 1
+
+        return count
+
+    @staticmethod
+    def total_months(datetime_obj):
+        """Method to return the total months."""
+        return datetime_obj.month + 12 * datetime_obj.year
+
+    def get_month_list(self):
+        """Gets the list of months that circuit maintenances have happened.
+
+        In order to know which months there are needed for a calculate average number of maintenances per month.
+
+        Returns:
+            list: List of months from first to last maintenance.
+        """
+        ordered_ckt_maintenance = CircuitMaintenance.objects.order_by("start_time")
+
+        # Check for a database with no maintenances in it, return nothing
+        if ordered_ckt_maintenance.first() is None:
+            return []
+
+        dates = [
+            str(ordered_ckt_maintenance.first().start_time.date()),
+            str(ordered_ckt_maintenance.last().start_time.date()),
+        ]
+        start, end = [datetime.datetime.strptime(_, "%Y-%m-%d") for _ in dates]
+        month_list = []
+        for tot_m in range(self.total_months(start) - 1, self.total_months(end)):
+            year, month = divmod(tot_m, 12)
+            month_list.append(datetime.datetime(year, month + 1, 1).strftime("%Y-%m"))
+        return month_list
+
+    def get_maintenances_per_month(self):
+        """Calculates the number of circuit maintenances per month.
+
+        Returns:
+            float: Average maintenances per month
+        """
+        # Initialize each month of maintenances
+        months = self.get_month_list()
```

Comment:
this feels like a lot more processing and work around the months than needed. Why not just something like:

```python
ordered_ckt_maintenance = CircuitMaintenance.objects.order_by("start_time")
if ordered_ckt_maintenance.count() < 2:
    return 0
delta = ordered_ckt_maintenance.last().start_time - ordered_ckt_maintenance.first().start_time
delta_months = delta.days / 365.0 * 12.0
if delta_months == 0:
    return 0
return self.queryset.count() / delta_months
```

Reply:
I have to investigate some more on this. This counts the days in between rather than the month itself. If there were only the months in the given year, this wouldn't be too bad. But then we get off a bit. The calculation gives a slightly off on the unit test.

Reply:
Found a way that will do this! Incoming.